packages/web/stores/index.tsx

```diff
@@ -1,11 +1,25 @@
 import React, { FunctionComponent, useState } from "react";
 
 import { RootStore } from "~/stores/root";
+import { api } from "~/utils/trpc";
 
 const storeContext = React.createContext<RootStore | null>(null);
 
 export const StoreProvider: FunctionComponent = ({ children }) => {
-  const [rootStore] = useState(() => new RootStore());
+  const apiUtils = api.useUtils();
+  const [rootStore] = useState(
+    () =>
+      new RootStore({
+        txEvents: {
+          onBroadcastFailed: () => {
+            apiUtils.edge.assets.getAsset.invalidate(); // Invalidate user balance
```

Comment:
Are we expected to remember to add all relevant queries here? IMO this may be unsustainable and error prone if we want to use tRPC for almost everything. I'm imagining a future with many bespoke queries.

To reopen the convo from just before christmas; perhaps we can use some sort of event emitter pattern (doesn't have to use event emitters since you said they're error prone) or similar to notify views what category of node data was affected by the tx. Then, we could invalidate queries closer to where they are used, which would help keep related code co-located. For common/foundational queries though, we could invalidate them in one place as we do here to avoid having to repeatedly invalidate for every query.

Reply:
It's unlikely that we'll need to monitor relevant queries for the global transaction events. This situation is unique because the user's balance needs to be invalidated, regardless of the transaction type.

As you mentioned, co-location seems to be a better strategy. My recommendation is to incorporate this within our upcoming useMutations. Relying on event-driven invalidation based on category is too prone to errors. It can lead to multiple invalidations across different components, potentially causing an n+1 refetching problem. This is because there's no mechanism to prevent repeated invalidations and they don't happen in a sequence.